tests/test_compute_utils.py

```diff
@@ -572,14 +577,147 @@ def test_surface_min_max(self):
+        for j in range(poly.shape[1]):
+            for k in range(poly.shape[2]):
+                for s in range(constant.size):
+                    a, b, c, d = poly[:, j, k]
+                    d = d - constant[s]
+                    np.testing.assert_allclose(
+                        roots[s, j, k],
+                        np.sort_complex(np.roots([a, b, c, d])),
+                    )
+
+    @pytest.mark.unit
+    def test_polyint(self):
+        """Test vectorized computation of polynomial primitive."""
+        quintic = 6
+        poly = np.arange(-90, 90).reshape(quintic, 3, -1) * np.e * np.pi
+        assert np.unique(poly.shape).size == poly.ndim
+        out = polyint(poly)
+        for j in range(poly.shape[1]):
+            for k in range(poly.shape[2]):
+                np.testing.assert_allclose(out[:, j, k], np.polyint(poly[:, j, k])[:-1])
+
+    @pytest.mark.unit
+    def test_polyder(self):
+        """Test vectorized computation of polynomial derivative."""
+        quintic = 6
+        poly = np.arange(-90, 90).reshape(quintic, 3, -1) * np.e * np.pi
+        assert np.unique(poly.shape).size == poly.ndim
+        out = polyder(poly)
+        for j in range(poly.shape[1]):
+            for k in range(poly.shape[2]):
+                np.testing.assert_allclose(out[:, j, k], np.polyder(poly[:, j, k]))
+
+    @pytest.mark.unit
+    def test_polyval(self):
+        """Test vectorized computation of polynomial evaluation."""
+        quintic = 6
+        poly = np.arange(-90, 90).reshape(quintic, 3, -1) * np.e * np.pi
+        assert np.unique(poly.shape).size == poly.ndim
+        x = np.linspace(0, 20, poly.shape[1] * poly.shape[2]).reshape(
+            poly.shape[1], poly.shape[2]
+        )
+        x = np.stack([x, x * 2], axis=-1)
+        x = np.stack([x, x * 2, x * 3, x * 4], axis=-1)
+        assert np.unique(x.shape).size == x.ndim
+        assert poly.shape[1:] == x.shape[: poly.ndim - 1]
+        assert np.unique((poly.shape[0],) + x.shape[poly.ndim - 1 :]).size == x.ndim - 1
+        val = polyval(x, poly)
+        for j in range(poly.shape[1]):
+            for k in range(poly.shape[2]):
+                np.testing.assert_allclose(val[j, k], np.poly1d(poly[:, j, k])(x[j, k]))
+
+    # TODO: FIXME
+    def bounce_point(
+        self, eq, lambdas, rho, alpha, max_bounce_points=20, max_field_line=10 * np.pi
+    ):
+        """Find bounce points."""
+        # TODO:
+        #     1. make another version of desc.backend.root_scalar
+        #        to avoid separate root finding routines in residual and jac
+        #        and use previous desc coords as initial guess for next iteration
+        #     2. write docstrings and use transforms in api instead of eq
+        def residual(zeta, i):
+            grid, data = field_line_to_desc_coords(rho, alpha, zeta, eq)
+            data = eq.compute(["|B|"], grid=grid, data=data)
+            return data["|B|"] - lambdas[i]
+
+        def jac(zeta):
```

Comment:
So for the rooting finding, you are using zeta as the field line following coordinate and for field line integration you are using theta_PEST?

Reply:
This should not cause a problem as long as, for a fixed alpha, theta_PEST and zeta are such that alpha stays the same.

Reply:
The only use of theta_PEST here should be to enable the conversion from field line coordinates (rho, alpha, zeta) to desc coordinates (rho, theta, zeta) in `field_line_to_desc_coords`.

For the root finding, we're doing a number of independent root finds. For every $\rho, \alpha$ field line, we find the roots of the function $f_{\rho, \alpha}(\zeta) = B_{\rho, \alpha}(\zeta) - 1/\lambda$, where $B_{\rho, \alpha}(\zeta) = |B|(\zeta)$ is the magnetic field along the field line following coordinate $\zeta$ of a particular field line labeled by $\rho, \alpha$. The result would give a bounce point for that value of $\lambda$ on that field line. For each field line we do a number of root finds each confined to particular interval to find more bounce points. (FYI, I switched to working on the spline method to compute the bounce integral before I finished a couple details of this one. I think it's better to make sure the spline methods work before giving more attention to this method).

Reply:
The integrals are over field line following zeta